.build/images/dietpi-imager

```diff
@@ -493,7 +502,34 @@
 			G_EXEC partprobe "$FP_SOURCE"
 			G_EXEC partx -u "$FP_SOURCE"
 		fi
-
+		if [[ "$ADD_DOS_PART" == "1" &&  "$PART_TABLE_TYPE" == 'dos' ]]
+		then
+			G_DIETPI-NOTIFY 2 'Adding a 4MB FAT partition to the DOS partition table'
+			G_EXEC_OUTPUT=1 G_EXEC dd if=/dev/zero bs=4M count=1 >> "$FP_SOURCE_IMG"
+			# first, methodically extract the starting point for that next partition
+			local old_table=$(sfdisk --json "$FP_SOURCE_IMG")
+			export lp_start=$(echo "$old_table" | jq ".partitiontable.partitions[].start" | sort -n | tail -1)
+			# jq syntax is a bit werid: pick the partition with the given start, and extract its size
+			local lp_size=$(echo "$old_table" | jq '.partitiontable.partitions|map(select(.start==(env.lp_start|tonumber)))[].size')
+			local np_start=$(( $lp_start + $lp_size ))
```

Comment:
See some lines below where we derive `$IMAGE_SIZE` from the last partition's end, which seems to be a simpler approach than parsing JSON. The first of these two lines without the single added sector (?) could be moved up, so the variable is available for this new block and can be increased by 4 MiB. Then the single sector and for GPT additional 33 sectors can be added below.

Reply:
Partitions can be out of order. But maybe we know that this isn't the case here since we created the partitions? I figured 'better safe than sorry' - but I'm happy to simplify this.
The single added sector is an old habit - there are some odd off by one errors that you can prevent from doing bad things to your data by having a spare sector between them.
Again, happy to remove that to be in line with your preferences.

Reply:
So what you do is scraping the partition with the highest start sector, regardless which partition number it has/at which position `sfdisk` lists them? Generally it makes sense, also since `dietpi-imager` is generally aimed to work as standalone script. But we do already assume that the root partition is the last partition, at least we minimise the last partition, and an out of order partition table might cause problems elsewhere already. If we want to deal with this properly, then I'd rather do this for all parts of the script, and also DietPi itself (the partition and filesystem expansion on first boot) instead of for this individual feature.

About the single added sector: I am just currently not sure whether `sfdisk` adds new partitions at including the start sector of from the given start sector on. With `parted` you can use the same sector as end sector for a previous and start sector for the next partition, but not sure whether this is the same for `sfdisk`. Generally I'd avoid a one sector gap. Actually I do not see where you add 1 sector in the new block? We do add it at the end of the whole image file after the last partition only (AFAIK there it was needed in some case), respectively 34 sectors for GPT for the backup partition table.

Reply:
> So what you do is scraping the partition with the highest start sector, regardless which partition number it has/at which position `sfdisk` lists them? 

Yes - as I mentioned elsewhere - I'm not a super bash programmer... but I love the swiss army knife tool chains out there. Like `sfdisk`/`json`/`jq`

> Generally it makes sense, also since `dietpi-imager` is generally aimed to work as standalone script. But we do already assume that the root partition is the last partition, at least we minimise the last partition, and an out of order partition table might cause problems elsewhere already. If we want to deal with this properly, then I'd rather do this for all parts of the script, and also DietPi itself (the partition and filesystem expansion on first boot) instead of for this individual feature.

I noticed that assumption that root is last... and ironically, I noticed that when I use this `add-fat` feature, I clearly break an assumption in the first boot and even though the fat partition is removed very early on, the root partition still didn't get resized in my testing...

I would love to see a more generic solution to this. If you give me some pointers where this should be implemented and what API you'd be looking for, I'd be happy to tackle that as well.

> About the single added sector: I am just currently not sure whether `sfdisk` adds new partitions at including the start sector of from the given start sector on. With `parted` you can use the same sector as end sector for a previous and start sector for the next partition, but not sure whether this is the same for `sfdisk`. Generally I'd avoid a one sector gap. Actually I do not see where you add 1 sector in the new block? We do add it at the end of the whole image file after the last partition only (AFAIK there it was needed in some case), respectively 34 sectors for GPT for the backup partition table.

I looked again and it turns out that I removed that. I am now using `local np_start=$(( $lp_start + $lp_size ))` - I originally had a `+ 1` in there, but apparently decided to remove that before submitting the PR 🙃
And yes - I see that you use "magic just add the partition" syntax elsewhere. I generally prefer to tell `sfdisk` exactly what I want and where I want it... fewer nasty surprises that way. But again - happy to follow your preferences...